Support mutable variables and immutable content

considerate opened the pull request:
This PR implements variable resources.

Note: The GC lock is a bit too aggressive, it hinders adding new content as well as modifying existing content. This is despite adding new content cannot make previously accessible things non-accessible. That said, the writes still happen, it's just that the moves are blocked until the GC has finished.

jonascarpay committed:
WIP experiment

jonascarpay committed:
Notes

jonascarpay committed:
BIG WIP

jonascarpay committed:
Weird types branch

jonascarpay committed:
WIP WIP

considerate committed:
var and ref (de)serialization

considerate committed:
remove stale comments

considerate committed:
drop Serialize super class for Content

considerate committed:
transact and borrow

considerate committed:
Expose the Store to users

considerate committed:
make U1 and V1 concrete in GRescope

jonascarpay committed:
WIP

jonascarpay committed:
Borrow in usage xample

jonascarpay committed:
WIP

jonascarpay committed:
Rescope store

considerate committed:
Add wrapper newtype for automatically encoding meta-data

jonascarpay committed:
drop the s

considerate committed:
types...

considerate committed:
WIP: implementation obligations everywhere

considerate committed:
WIP: open and write

considerate committed:
reading and writing of variables and references

considerate committed:
garbage collection

considerate committed:
set the GC Lock when sweeping

crvdgc commented:
- [ ] Tests should be updated.

crvdgc commented on src/Use.hs:
```diff
@@ -0,0 +1,81 @@
+  }
+
+-- >>> import qualified Data.Serialize as Serialize
+-- >>> let bytes = Serialize.runPut (Serialize.put exampleFoo)
+-- >>> bytes
+-- >>> let exampleFoo' = Serialize.runGet Serialize.get bytes :: Either String (Foo s)
+-- >>> let exampleFoo'' = fmap (\f -> f {val = 3}) exampleFoo'
+-- >>> Right exampleFoo == exampleFoo'
+-- >>> Right exampleFoo == exampleFoo''
+-- >>> error $ show exampleFoo'
+-- >>> error $ show exampleFoo
+-- "{\"_mli\":\"00000000-0000-0000-0000-000000000000\",\"lmi\":[\"00000000-0000-0000-0000-000000000000\",\"00000000-0000-0000-0000-000000000000\"],\"mi\":\"00000000-0000-0000-0000-000000000000\",\"rec\":\"00000000-0000-0000-0000-000000000000\",\"recs\":\"00000000-0000-0000-0000-000000000000\",\"val\":2}"
+-- True
+-- False
+-- Right (Foo {mi = 00000000-0000-0000-0000-000000000000, _mli = 00000000-0000-0000-0000-000000000000, lmi = [00000000-0000-0000-0000-000000000000,00000000-0000-0000-0000-000000000000], rec = 00000000-0000-0000-0000-000000000000, recs = 00000000-0000-0000-0000-000000000000, val = 2})
+-- Foo {mi = 00000000-0000-0000-0000-000000000000, _mli = 00000000-0000-0000-0000-000000000000, lmi = [00000000-0000-0000-0000-000000000000,00000000-0000-0000-0000-000000000000], rec = 00000000-0000-0000-0000-000000000000, recs = 00000000-0000-0000-0000-000000000000, val = 2}
+exampleFoo :: Foo s
+exampleFoo = Foo fakeVar fakeVar [fakeVar, fakeVar] fakeVar fakeVar 2
+
+-- >>> show fakeRef
+-- "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
+-- >>> import Ref(SHA)
+-- >>> import Var(UUID)
+-- >>> import Text.Read(readEither)
+-- >>> readEither (show fakeVar) :: Either String UUID
+-- >>> readEither (show fakeRef) :: Either String SHA
+-- Right 00000000-0000-0000-0000-000000000000
+-- Right AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
+someFunc :: IO Int
+someFunc =
+  openStore "/dev/null" (Root [] []) $ \(Root ls _) -> do
+    transact $
+      loop ls 0
+  where
+    loop [] z = pure z
+    loop (l : ls) z = do
+      foo1 <- readVar l
+      foo2 <- readVar (rec foo1)
+      writeVar (mi foo2) 0
+      loop ls (z + val foo1)
```
Should this be moved to tests?

considerate committed:
Have `a` in `retain` depend on the scope `s`

considerate committed:
potentially wait for GC for each resource

ramin-honary-xc committed:
write documentation for Casper module

ramin-honary-xc committed:
tests for 'openStore', 'transact', 'newVar', 'newRef', 'writeVar'

ramin-honary-xc committed:
Define instance for ByteString of Content typeclass

ramin-honary-xc committed:
Trouble dealing with forall'ed s in 'Casper.openStore' variable

ramin-honary-xc committed:
First test cases for new Casper with mutable 'Var' functionality

ramin-honary-xc committed:
test suite: a switch to choose between persistent or temporary store

ramin-honary-xc committed:
Bug fix: double encoding of Var content

ramin-honary-xc committed:
RawData, LazyRawData newtypes defined for unprefixed bytestrings

ramin-honary-xc committed:
Test round-trips on Refs

ramin-honary-xc committed:
pass root Var to openStore

ramin-honary-xc committed:
update comments, README.md

considerate committed:
Update comments

considerate committed:
Update comments

considerate committed:
Merge pull request #12 from xc-jp/ramin/define-tests

Define test cases

considerate committed:
comments

considerate committed:
update readme

considerate commented:
@jonascarpay I'll squash all of this into one or two commits since the history is quite messy. Is that OK with you?

jonascarpay commented:
@considerate I don't really care, nothing depends on the history being clean yet, and it might be nice to keep the blame intact, so I think you could make an argument here to keep it as is?

considerate commented:
@jonascarpay That's true for the time being but I'd like to merge this into master when I've verified that the API works with the rest of the system. The history is at least somewhat blocky so to retain the blame, squashing by author could be a tradeoff?

jonascarpay commented:
@considerate Even in master I'd be okay with a messy history, but looking through it yeah maybe by author is a good middle road

considerate committed:
perform the forkIO call within casper

considerate committed:
write new vars to the resource cache

considerate committed:
formatting

considerate committed:
Content for Maybe

considerate committed:
Content Char

considerate committed:
Drop the `s` again

I decided to drop the `s` parameter. It currently only ties variables
and references to a `Store`. However, that is marginal benefit.

The original intent of the parameter to tie it to a transaction is of
much higher value. Unfortunately, the transaction-scoped values have the
downside that it's difficult to created "reatained" variables in an
ergonomic way, where you can retain a variable and then rescope the
content to any arbitrary transaction.

The reason that is difficult is because then we'd need two kinds of
`RetainedVar`s one which adds an extra `s` parameter to its data type
when injected into a transaction, and another one that can work with
plain types like `Int`.

Let's punt on introducing more type safety in favor of better ergonomics
with regards to implementing systems on top of `casper`.

considerate committed:
Move modules into Data and Casper namespaces

considerate committed:
Content for more types

considerate committed:
re-export Serialize

considerate committed:
Content Void

considerate committed:
retain in same thread

considerate committed:
export retain

considerate committed:
Add MonadTrans instance to CasperT

considerate committed:
generalize `retain` to arbitrary `Content`

considerate committed:
Don't garbage collect in-use references

considerate committed:
drop Serialize constraint on Refs and Vars

considerate committed:
Orphan instance for Serialize Void

This is unfortunate but `cereal` doesn't provide this instance making
many downstream implementations unfeasible.

considerate committed:
do writes within the transaction STM block

considerate committed:
Collect non-garbage in GC

considerate committed:
Content (NonEmpty a) and Content Natural

considerate committed:
Expose reading references as ByteString

considerate committed:
Content instances for Identity and UUID

considerate committed:
lookups of untyped content with dirty checking

considerate committed:
Unsafe file paths

considerate committed:
MonadError instance

The instance requires UndecidableInstances because like the instance for
`MaybeT m`, `CasperT m` doesn't determine the error type `e`.

considerate committed:
expose `hoistCasper`

considerate committed:
expose lookupVar and lookupRef

considerate committed:
Only expose lookupRef due to safety issues with lookupVar
